Marcidy/testing errors

marcidy opened the pull request.

rizend commented on donate/routes.py:
```diff
@@ -143,8 +143,10 @@ def donation():
-        app.logger.error("StripeError: ")
-        flash(error)
+        err = error.json_body.get('error', {})
+        msg = err.get('message', {})
```
I would replace `err.get('message', {})` with `err.get('message', "")` since it looks to me like we want a string, not a dictionary to be used with `flash`.

rizend commented on donate/routes.py:
```diff
@@ -143,8 +143,10 @@ def donation():
             params['email'])
         app.logger.debug("Charge created: {}".format(charge_data))
     except StripeError as error:
```
using `StripeError` as a catch all without looking more carefully at what type of error it is is contrary to the Stripe api documentation's reccomendations and likely to result in these sorts of bugs in the future; the [handling errors](https://stripe.com/docs/api/errors/handling) Stripe docs layout what a better alternative looks like.

rizend commented on donate/routes.py:
```diff
@@ -143,8 +143,10 @@ def donation():
     except StripeError as error:
-        app.logger.error("StripeError: ")
-        flash(error)
+        err = error.json_body.get('error', {})
```
It is possible (and in fact the default case for the `StripeError` class, see [error.py, line 13](https://github.com/stripe/stripe-python/blob/f6adc3ea47ba9ebd5d57432c1ca33f0117661d6a/stripe/error.py#L13) and line 31) that `json_body` is `None` which will result in an error (`AttributeError: 'NoneType' object has no attribute 'get'`).  See previous comment for how to better handle errors.

marcidy commented:
@rizend I changed the existing StripeError a CardError to follow the pattern mentioned in the api doc.  I changed the gets to work sanely always.  err will always be a dict, msg will always be a string.
I added an explicilt ratelimit check, but everything else is absorbed in a StripeError because it will be logged and nothing should be sent to the user at that stage.

rizend commented on donate/routes.py:
```diff
@@ -142,9 +142,20 @@ def donation():
+
+    except se.CardError as error:
+        err = error.json_body.get('error', {})
+        msg = err.get('message', "")
```
s/`""`/`"Unknown Error"`/

rizend commented on donate/routes.py:
```diff
@@ -142,9 +142,24 @@ def donation():
+            app.logger.error("CardError: {}".format(error))
+        flash(msg)
+        return redirect('/index#form')
+    except se.RateLimitError as error:
```
Raised issue for logging, this will be seprate

rizend commented on donate/routes.py:
```diff
@@ -142,9 +142,24 @@ def donation():
+        return redirect('/index#form')
+    except se.StripeError as error:
+        app.logger.error("StripeError: {}".format(error))
+        flash("Unexpected error, please check data and try again."
```
Ideally add a space between the two sentences.
